contents/Documentation/FP concepts.playground/Pages/Functions vs Procedures.xcplaygroundpage/Contents.swift

```diff
@@ -4,11 +4,171 @@
+
+ Finally, let us look at the following function:
+ */
+func save(data: String) -> Bool {
+    print("Saving data: \(data)")
+    return true
+}
+/*:
+ We can see the function is total (defined for every input) and deterministic (always returns the same value for a given input). However, the execution of the function does something observable from the outside besides computing the output: it prints a message to the standard output. This makes this function **impure**.
+
+ ## Referential transparency
+
+ If a function is total, deterministic and pure, it has a property called **referential transparency**. An expression is referentially transparent if it can be replaced by the result of its evaluation and the behavior of the program does not change. This enables local equational reasoning. For instance, using the `add` function defined above, we can square the sum of two values:
+ */
+let square = add(x: 2, y: 5) * add(x: 2, y: 5)
+/*:
+ Since `add` is referentially transparent, we can replace calls to this function by the output it produces and the behavior of the program is not altered. This is known as common subexpression elimination.
+ */
+let sum = add(x: 2, y: 5)
+let square2 = sum * sum
+/*:
+ Notice that this is not necessarily the case when we have impure functions. Consider the following implementation of an impure `add` that logs the parameters it receives:
+ */
+func impureAdd(x: Int, y: Int) -> Int {
+    print("Received (\(x), \(y))")
+    return x + y
+}
+
+let impureSquare = impureAdd(x: 2, y: 5) * impureAdd(x: 2, y: 5)
+/*:
+ If we apply the same technique as above to eliminate the common subexpressions:
+ */
+let impureSum = impureAdd(x: 2, y: 5)
+let impureSquare2 = impureSum * impureSum
+/*:
+ The result of the two operations is the same, but the observed behavior of the overall program is not the same: the first implementation prints twice to the console, as there are two invocations to `impureAdd`, whereas in the second case there is only one print when the sum is saved in `impureSum`.
+
+ Referential transparency lets us have a better reasoning of our programs as we do not have to keep in mind the state of the computation and consider any side effects happening after the execution of a certain piece of code. Moreover, code that is referentially transparent is intrinsically testable (we just need to provide inputs and assert over the expected outputs) and can be optiimized and parallelized easily (there is no shared mutable state that causes concurrency issues).
+
+ ## Memoization
+
+ Another consequence of having a function that is referentially transparent is that it can be **memoized**. Memoization is a technique used to cache already computed values, specially if they have a high cost to be computed. When a function is memoized, you can think of it as a lookup table where inputs and their corresponding outputs are saved. Subsequent invocations just retrieve the value from the table instead of actually computing it.
```

Comment:
```suggestion
 Another consequence of having a function that is referentially transparent is that it can be **memoized**. Memoization is a technique used to cache already computed values, especially if they have a high cost to be computed. When a function is memoized, you can think of it as a lookup table where inputs and their corresponding outputs are saved. Subsequent invocations just retrieve the value from the table instead of actually computing it.
```